spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,118 @@
+package edu.umd.cs.findbugs.detect;
+
+import org.apache.bcel.Const;
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StringAnnotation;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DateFormatStringChecker extends OpcodeStackDetector {
+
+    private static final List<String> AM_PM_HOURS_FLAG_1 = new ArrayList<>(Arrays.asList("a", "h"));
+    private static final List<String> AM_PM_HOURS_FLAG_2 = new ArrayList<>(Arrays.asList("a", "K"));
+    private static final List<String> MILITARY_HOURS_FLAG_1 = new ArrayList<>(Arrays.asList(null,"H","a"));
+    private static final List<String> MILITARY_HOURS_FLAG_2 = new ArrayList<>(Arrays.asList(null,"k","a"));
+    private static final List<String> WEEK_YEAR_FLAG_1 = new ArrayList<>(Arrays.asList("w","Y","M","d"));
```

Comment:
In this list `Arrays.asList("w","Y","M","d")` the first element has different meaning. It would be better to introduce a class to replace this list.

```java
private static final class DateFormatRule {
  /** Required flag to run verification by this rule. Always apply rule if this field is {@code null}. */
  @Nullable
  String precondition;
  /** List of required flags in the pattern. All flags should exist in the given pattern. */
  @NonNull
  List<String> requiredFlags

  boolean verify(String dateFormatString) { ... }
}